### DISCUSSION TITLE
Implementing a client that connects to the serverr

### DISCUSSION THREAD

**Discussion by input54** [miguelgrinberg/python-socketio#1141] [Q&A]
I have both a webclient an webserver running with Python. The os I use is Windows 10
When I connecting to http://localhost:8080 I receive a 404:not found.

```
import socketio
sio = socketio.Client()
@sio.event
def connect():
    print (' connection established')
@sio.event
def disconnect():
    print(' disconnected from server')
sio.connect ('http://localhost:8080')
sio.emit('message', {'data':'my_data'})
sio.wait()
```
```
import socketio
from aiohttp import web
socket_io = socketio.AsyncServer()
app = web.Application()
socket_io.attach(app)

async def index(request):
    return web.Response(text='Hello world from socketio',content_type='text/html')

@socket_io.on('message')
def print_message(socket_id, data):
    print("Socket ID:' , socket_id")
    print("Data : ", data)
    app.router.add_get('/', index)

if __name__=='__main__':
    web.run_app(app)
```

**Comment by miguelgrinberg** [comment]
Your example works fine for me, so I don't know. The problem must be somewhere else. Maybe you have some other web server listening on port 8080 that is interfering with this one.

**Reply by input54** [reply]
Thank you miguelgrinberg
Do you also see a message in your browser localhost:8080.

**Reply by miguelgrinberg** [reply]
The app.router line must be in the global scope, you have it inside the function.

**Reply by input54** [reply]
I'm not sure yet. Below are the error messages I get.

```
Task exception was never retrieved
future: <Task finished name='Task-20' coro=<AsyncServer._handle_event_internal() done, defined at C:\Users\info\AppData\Local\Programs\Python\Python311\Lib\site-packages\socketio\asyncio_server.py:522> exception=RuntimeError('Cannot register a resource into frozen router.')>
Traceback (most recent call last):
  File "C:\Users\info\AppData\Local\Programs\Python\Python311\Lib\site-packages\socketio\asyncio_server.py", line 524, in _handle_event_internal
    r = await server._trigger_event(data[0], namespace, sid, *data[1:])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\info\AppData\Local\Programs\Python\Python311\Lib\site-packages\socketio\asyncio_server.py", line 562, in _trigger_event
    ret = handler(*args)
          ^^^^^^^^^^^^^^
  File "C:\Users\info\python\cybersecurity\Mastering-Python-for-Networking-and-Security-master\chapter1\code\web_server.py", line 14, in print_message
    app.router.add_get('/', index)
  File "C:\Users\info\AppData\Local\Programs\Python\Python311\Lib\site-packages\aiohttp\web_urldispatcher.py", line 1159, in add_get
    resource = self.add_resource(path, name=name)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\info\AppData\Local\Programs\Python\Python311\Lib\site-packages\aiohttp\web_urldispatcher.py", line 1085, in add_resource
    self.register_resource(resource)
  File "C:\Users\info\AppData\Local\Programs\Python\Python311\Lib\site-packages\aiohttp\web_urldispatcher.py", line 1047, in register_resource
    raise RuntimeError("Cannot register a resource into frozen router.")
RuntimeError: Cannot register a resource into frozen router.
```

**Reply by miguelgrinberg** [reply]
Are you familiar at all with aiohttp? This is an issue with that framework, and is not related to Socket.IO. The problem is likely that you are registering your route too late.

**Reply by input54** [reply]
No, I am not familiar with the module aiohttp. These scripts come from a Python book. I think I'll contact the publisher.
You don't know how to solve this problem either?

**Reply by miguelgrinberg** [reply]
Is it this? https://github.com/PacktPublishing/Mastering-Python-for-Networking-and-Security-Second-Edition/blob/144197e0dafc17454f977ba1efae382f133ec402/chapter2/code/socketio/web_socket_server.py

You may want to copy/paste the code to avoid copying mistakes, because that code looks correct to me. The code that you pasted above does not.

**Reply by input54** [reply]
Oke I give it a try. Thank you!
Yes it is working now.

**Reply by input54** [reply]
Hello miguelgrinberg,
I am very happy now that the scripts are working.
But can you tell me what is the different about the module socket and the module socketio and aiohttp.
For understanding I prefere the module socket.

Cheers,

Loek

**Reply by miguelgrinberg** [reply]
Are you asking about the `socket` module from the Python standard library? That is a low-level network socket. HTTP and Socket.IO are high-level network solutions that use the low-level network sockets in their implementation.

**Reply by input54** [reply]
Thank you. Yes I mean the socket module from the Python standard library. Do you mean that socket module is not secured ?
Is HTTP and Socket,.IO more secure ?

**Reply by miguelgrinberg** [reply]
No. HTTP and Socket.IO are high-level protocols. The socket module is low-level. It can only send and receive bytes.

**Reply by input54** [reply]
Hello miguelgrinberg,

As you know I am interesting in Python for Cybersecurity. Do you know the book Black Had Python and can you advise me for books about the topic Python for Cybersecurity.

Cheers,

Loek

**Reply by miguelgrinberg** [reply]
Sorry, I don't have any books to recommend on that topic.

**Reply by input54** [reply]
Hello miguelgrinbereg,
I have another quesstion for you and  I hope you can help me with it.
I bought the book Black Hat Python but in Chapter 2 with the script udp_client.py I receive an error in Windows 10.
I already sent an email to the publischer but the have not a dedicated Python programmer in the house so the advise me to search online for this error.

Below is the udp_clientscript:

```
import socket

HOST = '127.0.0.1'
PORT = 9997

client = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
client.sendto(b'AAABBBCCC', (HOST, PORT))

data, address = client.recvfrom(4096)
print(data.decode('utf-8'))
print(address.decode('utf-8'))

client.close()
```
```
Errorcode:

data, address = client.recvfrom(4096)
                    ^^^^^^^^^^^^^^^^^^^^^
ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host

```

**Reply by miguelgrinberg** [reply]
I'm sorry but I cannot answer your question, the topics you are learning have nothing to do with this project or the areas I'm experienced in.

**Reply by input54** [reply]
Oke thank you for the reply.
